doc/man/sieve-filter.1
.TH "SIEVE\-FILTER" "1" "11 July 2010"
.SH NAME
sieve\-filter \- Sieve mailbox filter tool for the Dovecot secure IMAP server
.PP
\fBWARNING: \fRThis tool is not finished and should \fB*NOT*\fR be used, unless you feel like testing newly developed
features! The behavior described in this manual page represents the design and not necessarily what the tool currently implements.

.SH SYNOPSIS
sieve\-filter [\fB\-c\fR \fIconfig\-file\fR] [\fIoptions\fR] \fIscript\-file\fR \fIsource\-location\fR \fIsource\-mailbox\fR [\fIinbox\-namespace\fR [\fInamespace\fR ...]]
.TP
\fInamepace\fR = [prefix=]location[;option=value,option=value,...]
.TP
[FIXME: what would be the easiest way to specify a filter operation without always needing to
delve into the complexity of namespaces]

.SH DESCRIPTION
.PP
The \fBsieve\-filter\fP command is part of the Pigeonhole Sieve implementation for the Dovecot secure 
IMAP server. Sieve (RFC 5228) is a simple and highly extensible language for filtering 
e\-mail messages. It can be implemented for any type of mail access protocol, mail 
architecture and operating system. The language cannot execute external programs and in 
its basic form it does not provide the means to cause infinite loops, making it suitable 
for running securely on mail servers where mail users have no permission run arbitrary programs.
.PP
The Sieve language was originally meant for filtering messages upon delivery. However, there are
occasions when it is desirable to filter messages that are already stored in a mailbox, for
instance when a bug in a Sieve script caused many messages to be delivered incorrectly.
Using the sieve\-filter tool it is possible to apply a Sieve script on all messages in a particular
mailbox, making it possible to delete messages, to store them in a different folder and to change
the assigned IMAP flags and keywords. Attempts to send messages to the outside world are ignored by default
for obvious reasons, but, using the proper command line options, it is possible to capture outgoing
mail as well. 
.PP
The command has three mandatory arguments: the \fIscript\-file\fP argument, which specifies the path of the
Sieve script, the \fIsource\-location\fP argument, which specifies the mail storage of the source mailbox 
(e.g. `maildir:~/Maildir'), and the \fIsoure\-mailbox\fP argument, which specifies the name of the source 
mailbox within the specified mail storage (e.g. `INBOX.Spam'). 
.PP
This tool does not (yet) use Dovecot's configuration file to obtain information on namespaces and the
location of mailboxes. Therefore, any used namespaces need to be specified on the command line. These
specifications directly follow the \fIsource\-mailbox\fP parameter. The first specified namespace will
be the INBOX namespace.
.PP
If no namespaces are defined on the commandline, the source\-location is used as the default mail store
where the INBOX is located. This means that the keep action could operate on the folder the message
originates from. In this case the message remains untouched and it is not duplicated, but IMAP flags and
keywords can be evaluated and changed with the imap4flags extension . If namespaces are defined explicitly, 
the source location is available as a namespace with prefix `#src/'. 
.PP
If no options are specified, the sieve\-filter command runs in a simulation mode in which it only 
prints what would be performed, without actually doing anything. Use the \fB\-e\fP option to activate
true script execution. Also, the source mailbox is opened read\-only by default, so that the source mailbox
remains unchanged. Use the \fB\-W\fP to allow changes in the source mailbox. 

.SH CAUTION
Although this is a very useful tool, it can also be very destructive when used improperly. A small 
bug in your Sieve script in combination with the wrong command line options could cause it to 
discard (many) more e\-mails than it was supposed to. Therefore, users are advised to read this manual
carefully and to use the simulation mode first to check what the script will do. 
.PP
\fBMAKING A BACKUP IS IMPERATIVE FOR ANY IMPORTANT MAIL!\fP
.PP
By default, it will open the source mailbox in a read\-only mode, such that it will not delete any of your
e\-mails. However, it can still litter other mailboxes with spurious copies of your e\-mails if your
Sieve script decides to do so.

.SH OPTIONS
.TP
\fB\-c\fP \fIconfig\-file\fP
Alternative Dovecot configuration file path.
.TP 
\fB\-D\fP \fIsource\-action\fP
By default, the sieve\-filter command does not delete the messages from the source mailbox. This means that
a copy operation is executed by default and the source mailbox is not altered. The \fIsource\-action\fP
parameter of the \fB\-D\fP option can take four different values:
.RS 7
.TP 
\fBkeep\fP (default)
Keep messages in source folder. If \fB\-W\fR is specified and the source mailbox is the destination of
a keep or fileinto action, flags can be changed by the Sieve script. Messages are never duplicated in the
source mailbox.
.TP 
\fBflag\fP
Flag messages as \\DELETED.
.TP 
\fBmove\fP [\fIfolder\fP]
Move messages to the indicated \fIfolder\fP.
.TP 
\fBexpunge\fP
Expunge messages, meaning that these are removed irreversibly when the tool finishes filtering.
.PP
Note that values other than `keep' have no effect, unless the \fB\-W\fP option is specified as well.
.RE
.TP
\fB\-e\fP
Turns on execution mode. By default, the sieve\-filter command runs in simulation mode in which it 
changes nothing, meaning that no mailbox is altered in any way and no actions are performed. It only
prints what would be done. Using this option the sieve\-filter command becomes active and performs the 
requested actions.
.TP
\fB\-f\fP \fIenvelope\-sender\fP
The envelope sender or return path. This is what Sieve's envelope test will compare to when the 
"from" envelope part is requested. Also, this is where response messages are sent to. 
.TP
\fB\-m\fP \fIdefault\-mailbox\fP
The mailbox within the default namespace where the keep action stores the message. This is "INBOX"
by default.
.TP
\fB\-Q\fP \fImail\-command\fP
Send outgoing e\-mail through the specified program. By default, the sieve\-filter command ignores 
Sieve actions such as redirect, reject, vacation and notify, but using this option outgoing messages
can be fed to the \fBstdin\fP of an external shell command. This option has no effect in simulation
mode, Unless you really know what you are doing, \fBDO NOT USE THIS TO FEED MAIL TO SENDMAIL!\f.
.TP
\fB\-r\fP \fIrecipient\-address\fP
The envelope recipient address. This is what Sieve's envelope test will compare to when the "to"
envelope part is requested. Some tests and actions will also use this as the owner's e\-mail address.
.TP
\fB\-S\fP \fIscript\-file\fP
Specify additional scripts to be executed before the main script. Multiple \fB\-s\fP arguments are
allowed and the specified scripts are executed sequentially in the order specified at the command
line.
.TP
\fB\-W\fP
Enables write access to the source mailbox. This allows deleting the messages from the source mailbox
and changing the assigned IMAP flags and keywords. 
.TP
\fB\-x\fP "\fIextension extension ...\fP"
Set the available extensions. The parameter is a space\-separated list of the active extensions. By
prepending the extension identifiers with \fB+\fP or \fB\-\fP, extensions can be included or excluded
relative to the default set of extensions. If no extensions have a \fB+\fP or \fB\-\fP prefix, only 
those extensions that are explicitly listed will be enabled. Unknown extensions are ignored and a 
warning is produced. By default, all supported extensions are available, except for deprecated extensions 
or those that are still under development.

For example \fB\-x\fP "+imapflags \-enotify" will enable the deprecated imapflags extension along with all
extensions that are available by default, except for the enotify extension.

.SH EXAMPLES

.TP
[...]

.SH AUTHOR
.PP
The Sieve implementation for Dovecot was written by Stephan Bosch <stephan@rename\-it.nl>.
.PP
Dovecot was written by Timo Sirainen <tss@iki.fi>.

.SH "SEE ALSO"
.BR sievec (1),
.BR sieve\-dump (1),
.BR sieve\-test (1)
.PP
Dovecot website: http://www.dovecot.org
.PP
Pigeonhole website: http://pigeonhole.dovecot.org


NEWS
v0.2.0 [TO BE RELEASED] Stephan Bosch <stephan@rename-it.nl>
	* Merged Sieve and ManageSieve packages into a single Pigeonhole package. 
	  There is also no need to patch Dovecot anymore to gain ManageSieve support.
	  Version numbering of previous Sieve releases is continued as v0.2.0. The
	  sources originally branched off from Sieve v0.1.5 and ManageSieve v0.11.4,
	  but the NEWS history of much more recent releases for Dovecot v1.2 is
	  included since these changes are all included in this release as well. 
	* The ManageSieve service now binds to TCP port 4190 by default due to the 
	  IANA port assignment for the ManageSieve service. When upgrading from v1.2,
	  this should be taken into account. The service can be configured manually to
	  listen on both 2000 and 4190.
	* The Dovecot configuration now calls the ManageSieve protocol 'sieve' in
	  stead of 'managesieve' because it is registered as such with IANA. The 
	  binaries and the services are still called managesieve and
	  managesieve-login.
	* The binary representation of a compiled Sieve script is updated to include
	  source code locations of all commands and arguments. This is implemented in
	  a similar manner as such debug information is included in some system
	  executables and libraries (DWARF-like). Run-time errors can now always refer 
	  to the proper line number in the Sieve source script.
	+ Simplified string matching API to use abstract string lists as data sources.
	  This will also make implementing the index extension easier in the future. 
	+ Significantly improved trace debugging with the sieve-test tool. The full
	  execution of the script can be examined, including the matched values and 
	  keys of the respective Sieve test commands. The executed statements are
	  listed with their line number (and code address when requested). The level
	  of detail is configurable from the command line. 
	+ The SIEVE and NOTIFY capabilities reported by the ManageSieve protocol can
	  now be configured manually. If left unconfigured, the capabilities are 
	  determined from the default Sieve and ManageSieve configuration.
	  User-specific capabilities aren't reported until after configuration.
	+ Significantly improved file error handling. This means that administrators
	  get a more useful and informative log message when file operations fail. The
	  most notable example is that when the LDA Sieve plugin is trying to store a
	  binary for a global script, the resulting failure message also points the
	  administrator towards pre-compiling the script with sievec.
	* Renamed the sieved tool to sieve-dump. The original name was somewhat
	  confusing.
	- Multiscript: fixed duplicate implicit keep caused by erroneous execution
	  state update.
	- Prevented assertion failure due to currupt binary string representation. 
	  If the string were missing a final \0 character an assertion was produced in
	  stead of a binary corruption error.
	- Imap4flags: fixed bug in setflag command; when parameter was a stringlist,
	  only the last item was actually set.
	- Variables extension: fixed :length set modifier to recognize utf8 characters
	  in stead of octets.
	- Testsuite: prevented innocent warning messages, i.e. part of the test,
	  from showing up by default.
	- ManageSieve/Sieve storage: fixed error handling of PUTSCRIPT commmand; save
	  commit errors would not make the command fail.

Sieve NEWS history:

v0.1.17 19-06-2010 Stephan Bosch <stephan@rename-it.nl>
	- Made sure source code positions for compiler messages are recorded at start
	  of tokens.
	- Fixed a few potential memory leaks in the Sieve compiler and the
	  spam/virustest extensions.
	- Made command line tools return proper exit status upon failure.

v0.1.16 30-04-2010 Stephan Bosch <stephan@rename-it.nl>

	* Finished implementation of spamtest, spamtestplus and virustest extensions.
	  These are not enabled by default and need to be activated with the
	  sieve_extensions setting. Documentation available in 
	  doc/spamtest-virustest.txt
	+ Vacation extension: the from address of the generated reply is now by 
	  default equal to whatever known recipient alias matched the headers of the
	  message. If it is one of the aliases specified with :addresses, it is used
	  in stead of the envelope recipient address that was used before. 
	+ Restructured and optimized the lexical scanner. 
	+ Added --with-docs configure option to allow disabling installation of 
	  documentation. 
	- Accidentally omitted 'extern' in two declarations of global variables in
	  header files, causing compile failures on certain systems.
	- Deprecated imapflags extension: fixed implicit assignment of flags. Turns
	  out this never really worked, but the effect of this bug was obscured by the
	  removeflag bug fixed in the previous release.
	- Fixed various memset argument mixups in enotify extension. This caused 
	  warnings on certain systems, but luckily no adverse effects at runtime.

v0.1.15 25-01-2010 Stephan Bosch <stephan@rename-it.nl>

	* Enotify extension: 
	  - Adjusted notify method API for addition of new notification methods.
	  - Set default importance level to 'normal' (was 'high').
	* Include extension: updated implementation towards most recent specification
	  (all should be backwards compatible):
	  - Implemented global variables namespace.
	  - Global command may now appear anywhere in a script.
	  - Implemented script name checking using the requirements specified in the
	    ManageSieve draft.
	  - One issue remains: ManageSieve currently requires included scripts to be
	    uploaded first, which is not according to specification.
	* Changed envelope path parser to allow to and from envelope addresses that
	  have no domain part. 
	+ Added preliminary support for Sieve plugins and added support for installing
	  Sieve development headers.
	+ Started work on the implementation of the spamtest, spamtestplus and
	  virustest extensions (unfinished).

src/sieve-tools/sieve-filter.c
/* Copyright (c) 2002-2010 Pigeonhole authors, see the included COPYING file
 */

#include "lib.h"
#include "lib-signals.h"
#include "ioloop.h"
#include "env-util.h"
#include "str.h"
#include "ostream.h"
#include "array.h"
#include "mail-namespace.h"
#include "mail-storage.h"
#include "mail-search-build.h"

#include "sieve.h"
#include "sieve-extensions.h"
#include "sieve-binary.h"

#include "sieve-tool.h"

#include "sieve-ext-debug.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <pwd.h>
#include <sysexits.h>

/*
 * Print help
 */

static void print_help(void)
{
	printf(
"Usage: sieve-filter [-m <mailbox>] [-x <extensions>] [-s <script-file>] [-c]\n"
"                    <script-file> <src-mail-store> [<dest-mail-store>]\n"
	);
}

enum discard_action_type {
	DISCARD_ACTION_KEEP,            /* Always keep messages in source folder */ 
	DISCARD_ACTION_DELETE,          /* Flag discarded messages as \DELETED */
	DISCARD_ACTION_TRASH_FOLDER,    /* Move discarded messages to Trash folder */      
	DISCARD_ACTION_EXPUNGE          /* Expunge discarded messages */
};

struct discard_action {
	enum discard_action_type type;
	const char *trash_folder;
};

static int filter_message
(struct mail *mail, struct sieve_binary *main_sbin, 
	struct sieve_script_env *senv, struct sieve_error_handler *ehandler,
	struct discard_action discard_action)
{
	struct sieve_exec_status estatus;
	struct sieve_binary *sbin;
	struct sieve_message_data msgdata;
	const char *recipient, *sender;
	int ret;

	sieve_tool_get_envelope_data(mail, &recipient, &sender);

	/* Initialize execution status */
	memset(&estatus, 0, sizeof(estatus));
	senv->exec_status = &estatus;

	/* Collect necessary message data */
	memset(&msgdata, 0, sizeof(msgdata));
	msgdata.mail = mail;
	msgdata.return_path = sender;
	msgdata.to_address = recipient;
	msgdata.auth_user = senv->username;
	(void)mail_get_first_header(mail, "Message-ID", &msgdata.id);

	/* Single script */
	sbin = main_sbin;
	main_sbin = NULL;

	/* Execute script */
	ret = sieve_execute(sbin, &msgdata, senv, ehandler, NULL);

	/* Handle message in source folder */
	if ( ret > 0 && !estatus.keep_original ) {
		switch ( discard_action.type ) {
		/* Leave it there */
		case DISCARD_ACTION_KEEP:
			sieve_info(ehandler, NULL, "message left in source folder");
			break;
		/* Flag message as \DELETED */
		case DISCARD_ACTION_DELETE:					
			sieve_info(ehandler, NULL, "message flagged as deleted in source folder");
			mail_update_flags(mail, MODIFY_ADD, MAIL_DELETED);
			break;
		/* Move message to Trash folder */
		case DISCARD_ACTION_TRASH_FOLDER:			
			sieve_info(ehandler, NULL, 
				"message in source folder moved to folder '%s'", 
				discard_action.trash_folder);
			break;
		/* Expunge the message immediately */
		case DISCARD_ACTION_EXPUNGE:
			sieve_info(ehandler, NULL, "message removed from source folder");
			mail_expunge(mail);
			break;
		/* Unknown */
		default:
			i_unreached();
			break;
		}
	}

	return ret;
}

/* FIXME: introduce this into Dovecot */
static void mail_search_build_add_flags
(struct mail_search_args *args, enum mail_flags flags, bool not)
{
	struct mail_search_arg *arg;

	arg = p_new(args->pool, struct mail_search_arg, 1);
	arg->type = SEARCH_FLAGS;
	arg->value.flags = flags;
	arg->not = not;

	arg->next = args->args;
	args->args = arg;
}

static int filter_mailbox
(struct mailbox *box, struct sieve_binary *main_sbin, 
	struct sieve_script_env *senv, struct sieve_error_handler *ehandler,
	struct discard_action discard_action)
{
	struct mail_search_args *search_args;
	struct mailbox_transaction_context *t;
	struct mail_search_context *search_ctx;
	struct mail *mail;
	int ret = 1;

	/* Sync mailbox */

	if ( mailbox_sync(box, MAILBOX_SYNC_FLAG_FULL_READ) < 0 ) {
		return -1;
	}

	/* Search non-deleted messages in the source folder */

	search_args = mail_search_build_init();
	mail_search_build_add_flags(search_args, MAIL_DELETED, TRUE);

	/* Iterate through all requested messages */

	t = mailbox_transaction_begin(box, 0);
	search_ctx = mailbox_search_init(t, search_args, NULL);
	mail_search_args_unref(&search_args);

	mail = mail_alloc(t, 0, NULL);
	while ( ret > 0 && mailbox_search_next(search_ctx, mail) > 0 ) {
		const char *subject, *date;
		uoff_t size = 0;
					
		/* Request message size */

		if ( mail_get_virtual_size(mail, &size) < 0 ) {
			if ( mail->expunged )
				continue;
			
			sieve_error(ehandler, NULL, "failed to obtain message size");
			continue;
		}

		if ( mail_get_first_header(mail, "date", &date) <= 0 )
			date = "";
		if ( mail_get_first_header(mail, "subject", &subject) <= 0 ) 
			subject = "";
		
		sieve_info(ehandler, NULL,
			"filtering: [%s; %"PRIuUOFF_T" bytes] %s", date, size, subject);
	
		ret = filter_message(mail, main_sbin, senv, ehandler, discard_action);
	}
	mail_free(&mail);
	
	/* Cleanup */

	if ( mailbox_search_deinit(&search_ctx) < 0 ) {
		ret = -1;
	}

	if ( ret < 0 ) {
		mailbox_transaction_rollback(&t);
		return -1;
	} else {
		if ( mailbox_transaction_commit(&t) < 0 ) {
			return -1;
		}
	}

	/* Sync mailbox */

	if ( mailbox_sync(box, MAILBOX_SYNC_FLAG_FULL_WRITE) < 0 ) {
		return -1;
	}
	
	return ret;
}

/*
 * Tool implementation
 */

int main(int argc, char **argv) 
{
	struct sieve_instance *svinst;
	const char *scriptfile, *recipient, *sender,
	*src_mailbox, *dst_mailbox, *src_mailstore, *dst_mailstore;
	struct mail_user *mail_user;
	struct mail_namespace *src_ns, *dst_ns;
	struct sieve_binary *main_sbin;
	struct sieve_script_env scriptenv;
	struct sieve_error_handler *ehandler;
	bool force_compile = FALSE;
	enum mailbox_flags open_flags =
		MAILBOX_FLAG_KEEP_RECENT | MAILBOX_FLAG_IGNORE_ACLS;
	enum mail_error error;
	struct discard_action discard_action = 
	{ DISCARD_ACTION_KEEP, "Trash" };
	struct mailbox *src_box;
	int c;

	sieve_tool = sieve_tool_init("sieve-filter", &argc, &argv, "m:x:P:CD", FALSE);

	/* Parse arguments */
	scriptfile = recipient = sender = src_mailstore = dst_mailstore = NULL;
	src_mailbox = dst_mailbox = "INBOX";
	force_compile = FALSE;
	while ((c = sieve_tool_getopt(sieve_tool)) > 0) {
		switch (c) {
		case 'm':
			/* default mailbox (keep box) */
			dst_mailbox = optarg;
			break;
		case 'C':
			force_compile = TRUE;
			break;
		default:
			print_help();
			i_fatal_status(EX_USAGE, "Unknown argument: %c", c);
			break;
		}
	}
	
	if ( optind < argc ) {
		scriptfile = t_strdup(argv[optind++]);
	} else {
		print_help();
		i_fatal_status(EX_USAGE, "Missing <script-file> argument");
	}

	if ( optind < argc ) {
		src_mailstore = t_strdup(argv[optind++]);
	} else {
		print_help();
		i_fatal_status(EX_USAGE, "Missing <mailstore> argument");
	}

	if ( optind < argc ) {
		dst_mailstore = t_strdup(argv[optind++]);
	}

	if ( optind != argc ) {
		print_help();
		i_fatal_status(EX_USAGE, "Unknown argument: %s", argv[optind]);
	}

	svinst = sieve_tool_init_finish(sieve_tool);

	/* Register Sieve debug extension */
	(void) sieve_extension_register(svinst, &debug_extension, TRUE);

	/* Create error handler */
	ehandler = sieve_stderr_ehandler_create(0);
	sieve_system_ehandler_set(ehandler);
	sieve_error_handler_accept_infolog(ehandler, TRUE);

	/* Compile main sieve script */
	if ( force_compile ) {
		main_sbin = sieve_tool_script_compile(svinst, scriptfile, NULL);
		(void) sieve_save(main_sbin, NULL);
	} else {
		main_sbin = sieve_tool_script_open(svinst, scriptfile);
	}

	sieve_tool_init_mail_user(sieve_tool, src_mailstore);
	mail_user = sieve_tool_get_mail_user(sieve_tool);

/*	if ( dst_mailstore != NULL ) {
		folder = "#src/";
		src_ns = mail_namespace_find(mail_user->namespaces, &folder);

		folder = "/";
		dst_ns = mail_namespace_find(mail_user->namespaces, &folder);	

		discard_action.type = DISCARD_ACTION_KEEP;	
	} else {*/
		dst_ns = src_ns = mail_user->namespaces;
		discard_action.type = DISCARD_ACTION_DELETE;	
/*	}*/

	/* Open the source mailbox */	
	src_box = mailbox_alloc(src_ns->list, src_mailbox, open_flags);
  if ( mailbox_open(src_box) < 0 ) {
		i_fatal("Couldn't open mailbox '%s': %s", 
			src_mailbox, mail_storage_get_last_error(src_ns->storage, &error));
  }

	/* Compose script environment */
	memset(&scriptenv, 0, sizeof(scriptenv));
	scriptenv.mailbox_autocreate = TRUE;
	scriptenv.default_mailbox = dst_mailbox;
	scriptenv.user = mail_user;
	scriptenv.username = sieve_tool_get_username(sieve_tool);
	scriptenv.hostname = "host.example.com";
	scriptenv.postmaster_address = "postmaster@example.com";
	scriptenv.smtp_open = NULL;
	scriptenv.smtp_close = NULL;
	scriptenv.duplicate_mark = NULL;
	scriptenv.duplicate_check = NULL;
	scriptenv.trace_stream = NULL;

	/* Apply Sieve filter to all messages found */
	(void) filter_mailbox
		(src_box, main_sbin, &scriptenv, ehandler, discard_action);
	
	/* Close the mailbox */
	if ( src_box != NULL )
		mailbox_free(&src_box);

	/* Cleanup error handler */
	sieve_error_handler_unref(&ehandler);
	sieve_system_ehandler_reset();

	sieve_tool_deinit(&sieve_tool);

	return 0;
}
